californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -310,11 +317,19 @@ public void completed(final Exchange exchange) {
 				} else {
 					// in case an empty ACK was lost
 					KeyMID idByMID = KeyMID.fromOutboundMessage(originRequest);
-					exchangeStore.remove(idByMID);
+					exchangeStore.remove(idByMID, exchange);
 				}
 				KeyToken idByToken = KeyToken.fromOutboundMessage(originRequest);
-				exchangeStore.remove(idByToken);
-				if(!originRequest.getOptions().hasObserve()) {
+				exchangeStore.remove(idByToken, exchange);
+				Request request = exchange.getRequest();
+				if (request != originRequest && null != request &&  null != request.getToken()
+						&& !Arrays.equals(request.getToken(), originRequest.getToken())) {
+					// remove starting request also
+					originRequest = request;
+					idByToken = KeyToken.fromOutboundMessage(originRequest);
+					exchangeStore.remove(idByToken, exchange);
```

Comment:
About the confusion I totally agree :-).

See also #232. There we search for observes in the exchangeStore, where we have already removed those Observe-Exchanges with the "onNotification" redesign :-).

I feel , that we now try to cleanup in a lot of places. But I don't feel, that we ensure not to forgett something. And removing the token or MID for a Exchange multipe times may be dangerous, if the token or MID is reused in the meantime. The most important of that PR is therefore the "save remove" of Exchanges. And checking for "leaks", I found the stuff wiht the two request an dadded the cleanup in the ExcahngeObserver, mainly because I don't feel, that it is easy to ensure, that the blockwise layer will correct the currentRequest before completion.

If the blockwise should do the "special request" cleanup, the blockwise layer should be informed about the completion. So one way to do that would introduce a special ExchangeObserver for (blockwise)Exchanges, set in the blockwise layer into the Exchange. But I don't think this would pay off.